libraries/adaptive-expressions/src/triggerTrees/node.ts

```diff
@@ -0,0 +1,362 @@
+/**
+ * @module adaptive-expressions
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Clause } from './clause';
+import { RelationshipType } from './relationshipType';
+import { TriggerTree } from './triggerTree';
+import { Trigger } from './trigger';
+
+enum Operation {
+    none = 'none',
+    found = 'found',
+    added = 'added',
+    removed = 'removed',
+    inserted = 'inserted'
+}
+
+/**
+ * Node in a trigger tree.
+ */
+export class Node {
+    private _allTriggers: Trigger[] = [];
+    private _triggers: Trigger[] = [];
+    private _specializations: Node[] = [];
+
+    /**
+     * Intializes a new instance of the `Node` class.
+     * @param clause The logical conjunction this node represents.
+     * @param tree The trigger tree this node is found in.
+     * @param trigger The trigger to initialize this node.
+     */
+    public constructor(clause: Clause, tree: TriggerTree, trigger?: Trigger) {
+        this.clause = new Clause(clause);
+        this.tree = tree;
+        if (trigger) {
+            this._allTriggers.push(trigger);
+            this._triggers.push(trigger);
+        }
+    }
+
+    /**
+     * Gets all of the most specific triggers that contains the `Clause` in this node.
+     */
+    public get triggers(): Trigger[] {
+        return this._triggers;
+    }
+
+    /**
+     * Gets all triggers that contain the `Clause` in this node.
+     */
+    public get allTriggers(): Trigger[] {
+        return this._allTriggers;
+    }
+
+    /**
+     * Gets specialized children of this node.
+     */
+    public get specializations(): Node[] {
+        return this._specializations;
+    }
+
+    /**
+     * Gets the logical conjunction this node represents.
+     */
+    public clause: Clause;
+
+    /**
+     * Gets the tree this node is found in.
+     */
+    public tree: TriggerTree;
+
+    /**
+     * Gets a string that represents the current node.
+     * @param builder An array of string to build the string of node.
+     * @param indent An integer representing the number of spaces at the start of a line.
+     */
+    public toString(builder: string[] = [], indent = 0): string {
+        return this.clause.toString(builder, indent);
+    }
+
+    /**
+     * Identify the relationship between two nodes.
+     * @param other Node to compare against.
+     * @returns Relationship between this node an the other.
+     */
+    public relationship(other: Node): RelationshipType {
+        return this.clause.relationship(other.clause, this.tree.comparers);
+    }
+
+    /**
+     * Gets the most specific matches below this node.
+     * @param state Frame to evaluate against.
+     * @returns List of the most specific matches found.
+     */
+    public matches(state: any): Trigger[] {
```

Comment:
Same issue WRT `state` being `any`.

Reply:
Changed type of `state` to `MemoryInterface | any` to match the `tryEvaluate()` method.